alibi/explainers/anchor_tabular.py

```diff
@@ -48,224 +74,273 @@ def fit(self, train_data: np.ndarray, disc_perc: list = [25, 50, 75]) -> None:
+            return
+
+        first_numerical_idx = np.searchsorted(self.categorical_features, self.numerical_features[0]).item()
+        if first_numerical_idx > 0:  # First column(s) might contain categorical data
+            for cat_enc_idx in range(0, first_numerical_idx):
+                self.cat_lookup[cat_enc_idx] = X[cat_enc_idx]
+                self.enc2feat_idx[cat_enc_idx] = cat_enc_idx
+
+        ord_enc_idx = first_numerical_idx - 1  # -1 as increment comes first
+        for i, feature in enumerate(self.numerical_features):
+            n_bins = len(self.feature_values[feature])
+            for bin_val in range(n_bins):
+                ord_enc_idx += 1
+                self.enc2feat_idx[ord_enc_idx] = feature
+                # if feat. value falls in same or lower bin, sample from same or lower bin only ...
+                if X[feature] <= bin_val != n_bins - 1:
+                    self.ord_lookup[ord_enc_idx] = set(i for i in range(bin_val + 1))
+                # if feat. value falls in a higher bin, sample from higher bins only
+                elif X[feature] > bin_val:
+                    self.ord_lookup[ord_enc_idx] = set(i for i in range(bin_val + 1, n_bins))
+                else:
+                    del self.enc2feat_idx[ord_enc_idx]
+                    ord_enc_idx -= 1  # when a discretized feat. of the instance to be explained falls in the last bin
+
+            # check if a categorical feature follows the current numerical feature & update mappings
+            if i < len(self.numerical_features) - 1:
+                n_categoricals = self.numerical_features[i + 1] - self.numerical_features[i] - 1
+                if n_categoricals > 0:
+                    cat_feat_idx = feature + 1
+                    for cat_enc_idx in range(ord_enc_idx + 1, ord_enc_idx + 1 + n_categoricals):
+                        self.cat_lookup[cat_enc_idx] = X[cat_feat_idx]
+                        self.enc2feat_idx[cat_enc_idx] = cat_feat_idx
+                        cat_feat_idx += 1
+                    ord_enc_idx += n_categoricals
+
+        # check if the last columns are categorical variables and update mappings
+        last_num_idx = np.searchsorted(self.categorical_features, self.numerical_features[-1]).item()
+        if last_num_idx != len(self.categorical_features):
+            cat_enc_idx = max(self.ord_lookup.keys()) + 1
+            for cat_feat_idx in range(self.numerical_features[-1] + 1, self.categorical_features[-1] + 1):
+                self.cat_lookup[cat_enc_idx] = X[cat_feat_idx]
+                self.enc2feat_idx[cat_enc_idx] = cat_feat_idx
+                cat_enc_idx += 1
+
+    def sample_from_train(self, anchor: list, val2idx: dict, ord_lookup: dict, cat_lookup: dict, enc2feat_idx: dict,
+                          num_samples: int) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Sample data from training set but keep features which are anchor in the proposed anchor the same
```

Comment:
> which are anchor in the proposed anchor

which are present?

Reply:
Great spot, will update! It made no sense whatsoever before!